### ISSUE TITLE
[Request] LSDJ compatibility.

### ISSUE THREAD

**Issue by CastleTiles** [chrismaltby/gb-studio#30]
Is there any way you can add or implement LSDJ .sav files for soundtracks?

**Comment by thomasmaclean** [comment]
Hey there... while the file format is understood, there isn't really an open-source player for the lsdsng format other than the one embedded in the LSDJ ROM. As such, I don't think something like this would be possible (although I'd love it if it was). Although mod format isn't nearly as powerful, it does allow you to get the basics down.

**Comment by clsource** [comment]
I think this could be implemented by using VGM as requested in #44 

Steps 
1 - Load sav file in Little Sound DJ
2 - Create a VGM file by using a similar method as described here

https://project2612.org/tutorial/02-logging.htm

This is just a theory but would be nice to have.

**Comment by SeeBeyondMusic** [comment]
This library also exists https://github.com/stijnfrishert/liblsdj

**Comment by clsource** [comment]
@SeeBeyondMusic the problem with that lib is that only works as a export-import tool. You can easily manipulate sav files but cannot reproduce those in a stand alone game. Another approach is using that lib and somehow convert the lsdsng file to gbt. In that case you could use that.

**Comment by Pigu-A** [comment]
VGM logs (which liblsdj is basically generating) would run the ROM space out really fast though. Which is the only reason why I don't like this method.

**Comment by cyberic99** [comment]
> @SeeBeyondMusic the problem with that lib is that only works as a export-import tool. You can easily manipulate sav files but cannot reproduce those in a stand alone game. Another approach is using that lib and somehow convert the lsdsng file to gbt. In that case you could use that.

A way to get register dumps would be:
https://github.com/jkotlinski/lsdpack

But as @Pigu-A says, it could use a lot of ROM space

Maybe it could be possible to implement a kind of compression

In addition, LSDj song can be very CPU-extensive, leaving only a few CPU cycles for the main application...

**Comment by jkotlinski** [comment]
Hi!
To add some facts to the discussion.

- lsdpack player does not take an extreme amount of CPU. I don't have exact measurements but I would estimate maybe 10% CPU. Far less than LSDj itself.
- lsdpack takes plenty of ROM space, but since MBC5 allows for 8 MByte ROMs, I found space is not that much of a concern. It has already been used to make full-length albums. The ROM usage is also highly dependent on what features are used in the songs. Avoiding sample playback and high-speed vibrato/pitch bend helps.

**Comment by Pigu-A** [comment]
I have inspected lsdpack's code and noticed that it already used a simple kind of compression. So I'm fine with that enough as a register log solution. Although I can think of a better compression scheme that doesn't take that much CPU time. Maybe I'll write up a format proposal and contribute there.

**Comment by jkotlinski** [comment]
Suggestions would be gladly accepted!

**Comment by PixelShift-gb** [comment]
I emailed them and they said you can export music using this link
https://github.com/jkotlinski/lsdpack

**Comment by jkotlinski** [comment]
Is there anyone who would be interested in integrating this? I am wondering what I can do to help from my side.

**Comment by jkotlinski** [comment]
lsdpack has now been updated to generate .gbs files. I hope this in some way can help GB Studio integration.

https://github.com/jkotlinski/lsdpack

**Comment by RichardULZ** [comment]
.gbs files are of no help, the Game boy soundsystem format is often even harder to integrate into a game!

The gbstudio engine rewrite (look for the gbvm repository) is being designed to allow more external data, specifically hUGETracker.
toxa has figured out how to pack and link hUGE's rgbds code with the gbdk engines, and data for the game compile will be more human readable, so adding data not from gbstudio will be a little easier.

The lsdpack driver above requires firing 360 times per second! 
The new hUGE is being configured with the interupt timer, to run main code every 64hz, but also at 256hz, for potential wave channel sample playback (wave samples use a full 16kb bank in 4 seconds btw).

So, whenever the engine rewrite and importantly editor overhaul to support this new engine is near, lsdpack would be closer to a possibility, but still has its own hurdles.

**Comment by jkotlinski** [comment]
Hi Richard!
Thanks for your reply!

I am not sure if there was a question, but since you seem interested in the
matter, I can expand on how often the lsdpack driver needs to be called (I
should update the README with this info)

 * fullspeed - call 6 times/frame - 11468 Hz samplerate. This is the
default.
 * halfspeed - call 3 times/frame - 5734 Hz samplerate. This saves ROM+CPU
but still allows some samples.
* CPU saving - if no sample playback is needed, its perfectly fine to call
the driver only once per frame, relying on VBL interrupt only.(With current
implementation, driver would also need five dummy calls to maintain tempo,
but that’s just a detail that could be adjustedj

Cheers /J

**Comment by RichardULZ** [comment]
Neat, @untoxa is the one integrating hUGEtracker for the next version, 
They just mentioned on the discord they'd considered the idea of "pluggable music drivers", though, and would love help writing wrappers. 
They might be able to teach you the secrets of Rgbds to Gbdk linking.

**Comment by jkotlinski** [comment]
@untoxa is this correct? can the secrets be taught?

**Comment by untoxa** [comment]
What secrets?

**Comment by jkotlinski** [comment]
RichardULZ commented 7 days ago

> 
> 
> Neat, @untoxa is the one integrating hUGEtracker for the next version,
> They just mentioned on the discord they'd considered the idea of "pluggable music drivers", though, and would love help writing wrappers.
> They might be able to teach you the secrets of Rgbds to Gbdk linking.

**Comment by untoxa** [comment]
Linking rgbds objects? Readme on the front page of this repo: https://github.com/untoxa/hUGEBuild

but that is hardly needed for that lsdj player which code is very primitive, you can simply rewrite it.

**Comment by thomasmaclean** [comment]
@chrismaltby Wait... was this actually completed? I was cruising through the commit history and I didn't see any code changes that looked like they were related to this. Is this more of a won't fix?

**Comment by chrismaltby** [comment]
Sorry yes, was trying to clean up the 100s of old issues. I mean't to set this one as not planned :-)

**Comment by thomasmaclean** [comment]
Thanks for the clarification!

**Comment by chromix** [comment]
In case anyone finds this issue again in the future, you can now export UGE files from LSDJ SAV files using this utility:
https://beatscribe.itch.io/lsdj-to-uge

Rather than do instrument conversions it requires you to name your LSDJ instruments a certain way which map to a set of UGE instrument templates. Even so, just getting the pattern/table/song data into UGE would probably be a big leg-up for most folks looking to do this kind of conversion, and it makes it possible to use LSDJ in your creative workflow.
